File: CS21M070_L10/HuffManTreeADT.py
# ADT of huffman tree
class HuffManTree:
    
    # constructor
    def __init__(self, symbol=None, probability=None, lChild=None, rChild=None):
        self.symbol = symbol    # set all required parameters
        self.probability = probability
        self.lChild = lChild
        self.rChild = rChild
    
    # overload str method to print the class using 'print' 
    def __str__(self):
        return f"({self.symbol} | {self.probability})" # f-strings are faster and more efficient
        
    def getSymbol(self):
        return self.symbol

    def getProbability(self):
        return self.probability
    
    def setSymbol(self, symbol):
        self.symbol = symbol
    
    def setProbability(self, probability):
        self.probability = probability
        
    def setLeftTree(self, lChild):
        self.lChild = lChild
        
    def setRightTree(self, rChild):
        self.rChild = rChild
        
    def getLeftTree(self):
        return self.lChild

    def getRightTree(self):
        return self.rChild
    
    #---------------------------------------------------------------------------#
    # printCodes -- does a pre-order traversal to print the codes from the tree
    # Input: an auxillary stack to keep track of the path followed to reach a node
    # Output: NA, just prints the codes
    # Bugs: NA
    #---------------------------------------------------------------------------#
    def printCodes(self, auxStack):       
        if self.symbol != 'I':  # if we have hit a leaf node, not intermediate one
            print(self.symbol, end=' ')  # print the node content
            print(''.join(auxStack))   # print the stack which contains the path we have followed to reach here

        if self.lChild is not None: # if we are at an intermediate node, check if left subtree is present, if yes
            auxStack.append('1')    # push a 1 to the top of the stack to indicate going left
            self.lChild.printCodes(auxStack) # now recurse on the left substree
            temp = auxStack.pop()   # once back, pop the stack top
            
        if self.rChild is not None:     # if we are at an intermediate node, check if right subtree is present, if yes#
            auxStack.append('0') # push a 0 to the top of the stack to indicate going right
            self.rChild.printCodes(auxStack) # now recurse on the right substree
            temp = auxStack.pop()   # once back, pop the stack top

File: CS21M070_L8/src/HuffManTreeADT.cpp
#include "../include/HuffManTreeADT.h"
#include <iostream>
using namespace std;

HuffManTree::HuffManTree() {}

HuffManTree::HuffManTree(float prob, char letter)
{
    symbol = 'I';
    probability = 0;
    lChild = NULL;
    rChild = NULL;
}

HuffManTree::HuffManTree (float prob, char letter, HuffManTree *lChild, HuffManTree *rChild)
{
    probability = prob;
    symbol = letter;
    lChild = NULL;
    rChild = NULL;
}

void HuffManTree::createBinaryTree(float prob, char letter)
{
    symbol = letter;
    probability = prob;
}

char HuffManTree::getSymbol()
{
    return symbol;
}

float HuffManTree::getProbability()
{
    return probability;
}

void HuffManTree::setSymbol(char sym)
{
    symbol = sym;
}

void HuffManTree::setProbability(float prob)
{
    probability = prob;   
}

HuffManTree* HuffManTree::getLeftTree()
{
    return lChild;
}

HuffManTree* HuffManTree::getRightTree()
{
    return rChild;
}

void HuffManTree::setLeftTree(HuffManTree *ltree)
{
    lChild = ltree;
}

void HuffManTree::setRightTree(HuffManTree *rtree)
{
    rChild = rtree;
}

void HuffManTree::makeBinaryTree(HuffManTree *ltree, HuffManTree *rtree)
{
    setLeftTree(ltree);
    setRightTree(rtree);
}

/*********************************************************************
 * printCodes -- printes the formed binary tree in pre-order manner
 * Input: The subtree root as a pointer
 * Output: NA, just prints the tree in pre-order manner
 * Bugs: NA
 *********************************************************************/
void HuffManTree::printCodes(HuffManTree *tree)
{
    static char code[50];   /*Helps as a stack to keep track of movement, i.e. left or right*/
    static int index = 0;   /*Helps keep track of the stack top*/
    
    HuffManTree* left = tree -> getLeftTree();  /*Store the left subtree, right subtree pointers to access later*/
    HuffManTree* right = tree -> getRightTree();
    char data = tree -> getSymbol();    /*Store the character in the node to print later*/
    
    if(data != 'I') /*If the node is not an internal node, i.e. it is a leaf node*/
    {
        cout << data << ' ';    /*print the node character*/
        for(int i = 0; i < index; i++)
            cout << code[i];    /*print the stack which is in the form of an array, i.e. the path to get this leaf*/
        cout << '\n';
    }
    if(left)    /*if the left node is available*/
    {
        code[index++] = '1';    /*put a 1 on top of the stack*/
        printCodes(left);       /*and move to the left*/
        index--;                /*decrease the stack top, to pop-off the previous element*/
    }
    if(right)   /*if the right node is also available*/
    {
        code[index++] = '0';    /*put a 0 on top of the stack*/
        printCodes(right);      /*and move to the right*/
        index--;                /*decrease the stack top, to pop-off the previous element*/
    }
}